inkcpp/globals_impl.cpp

```diff
@@ -68,6 +68,80 @@ namespace ink::runtime::internal
 		return _variables.get(name);
 	}
 
+	value* globals_impl::get_variable(hash_t name) {
+		return _variables.get(name);
+	}
+
+	template<typename T, const T* (value::*FN)() const, typename ... TYPES>
+	auto fetch_variable( const value* v, TYPES ... types) {
+		return v && ((v->get_data_type() == types) || ...)
+			? (v->*FN)()
+			: nullptr;
+	}
+	template<typename T, T* (value::*FN)(), typename ... TYPES>
+	auto fetch_variable(value* v, TYPES ... types) {
+		return v && ((v->get_data_type() == types) || ...)
+			? (v->*FN)()
+			: nullptr;
+	}
+
+	const uint32_t* globals_impl::get_uint(hash_t name) const {
+		return fetch_variable<uint32_t, &value::as_uint_ptr>(get_variable(name), data_type::uint32);
+	}
+	bool globals_impl::set_uint(hash_t name, uint32_t val) {
+		uint32_t* p = fetch_variable<uint32_t, &value::as_uint_ptr>(get_variable(name), data_type::uint32);
+		if (p == nullptr) { return false; }
+		*p = val;
+		return true;
+	}
+
+	const int32_t* globals_impl::get_int(hash_t name) const {
+		return fetch_variable<int32_t, &value::as_int_ptr>(get_variable(name), data_type::int32);
+	}
+	bool globals_impl::set_int(hash_t name, int32_t val) {
+		int32_t* p = fetch_variable<int32_t, &value::as_int_ptr>(get_variable(name), data_type::int32);
+		if (p == nullptr) { return false; }
+		*p = val;
+		return true;
+	}
+
+	const float* globals_impl::get_float(hash_t name) const {
+		return fetch_variable<float, &value::as_float_ptr>(get_variable(name), data_type::float32);
+	}
+	bool globals_impl::set_float(hash_t name, float val) {
+		float* p = fetch_variable<float, &value::as_float_ptr>(get_variable(name), data_type::float32);
+		if (p == nullptr) { return false; }
+		*p = val;
+		return true;
+	}
+
+	const char * const * globals_impl::get_str(hash_t name) const {
+		return fetch_variable<const char*, &value::as_str_ptr>(
```

Comment:
This will fail if the value is a string composed of multiple concatenated data types that aren't strings. Example `5 + "hi" + 4`. `get_data_type()` just returns the type of the _first_ piece of data in the value. It's `type()` that does the resolution to realize that a value with a mix of numbers and strings is actually a string.

Reply:
Oh, yes, this is a thing, because:
The getter is const, but modifying a value is not, maybe we can make the `_variables` mutable? I don't realy see a other way to solve this problem,
creating each time a new string table entry and write it is also possible, but then we must parse it every time ...

also the get<const char*> should contain a warning that the value is only valid until the next runner call ^^